assets/js/main.js

```diff
@@ -0,0 +1,250 @@
+        // colliding with the pad
+        pos.x >= padPos.x && pos.x <= padPos.x + game.padWidth &&
+        // out of bounds
+        pos.y === padPos.y) {
+        onDropCatch($drop);
+      }
+    }
+
+    // drop removeal is separated to avoid removing from an array midway
+    // through iterating
+    for (var i = 0; i < $missed.length; i++) {
+      var $drop = $missed[i];
+      onDropMiss($drop);
+    }
+  };
+
+  // ================== pad ==================
+  // pad related functions
+
+  var movePad = function () {
+    var pos = getXY($pad);
+    var newPos = pos.x + game.padDX;
+    // don't move the pad if it's out of bounds
+    if (newPos + game.padWidth <= GRID_WIDTH && newPos >= 0) {
+      move($pad, newPos, pos.y);
+    }
+  };
+
+  // ================== actions ==================
+  // game actions
+
+  var onDropMiss = function ($drop) {
+    removeDrop($drop);
+  };
+
+  var onDropCatch = function ($drop) {
+    playSound(catchSound);
+    game.score = game.score + 1;
+    $score.html(game.score);
+    removeDrop($drop);
+  };
+
+  // NOTE: we must use key down and up as keypress doesn't work for arrow keys
+  // set the pads movement direction based on key down
+  var onKeyDown = function (event) {
+    if (event.keyCode === LEFT) {
+      game.padDX = -1;
+    } else if (event.keyCode === RIGHT) {
+      game.padDX = 1;
+    }
+  };
+
+  // when the user lets go of the key set the pads movement to 0
+  var onKeyUp = function (event) {
+    game.padDX = 0;
+  };
+
+  // ================== game ==================
+  // high level game functions
+
+  // restart the game this end and restarts the game - not currently used
+  var restart = function () {
+    // cleanup if there was a last game
+    if (game && game.tickHandle) {
+      clearInterval(game.tickHandle);
+    }
+    for (var i = 0; i < $drops.length; i++) {
+      $drops[i].remove();
+    }
+    // start a new game
+    $drops = [];
+    game = {
+      score: 0,
+      tick: START_TICK,
+      padWidth: START_PAD_WIDTH,
+      dropCreateTime: START_DROP_CREATE_TIME,
+      dropsSpeed: START_DROPS_SPEED,
+      moveSpeed: START_MOVE_SPEED,
+      time: 0
+    };
+    $score.html(game.score);
+    size($pad, game.padWidth, 1);
+    // start the game pad in the middle of the board at the bottom
+    move($pad, Math.round(GRID_WIDTH / 2), GRID_HEIGHT - 2);
+    game.tickHandle = setInterval(update, game.tick);
+  };
+
+  // setup the game this is called only once per page load
+  var setup = function () {
+    $game = $('.game');
+    $game.css({
+      width: GRID_WIDTH * GRID_SIZE,
+      height: GRID_HEIGHT * GRID_SIZE
+    });
+    $pad = $('.pad');
+    $score = $('.score');
+    $(document).keydown(onKeyDown);
+    $(document).keyup(onKeyUp);
+    catchSound = new Audio('assets/sound/catch.mp3');
+  };
+
+  var update = function () {
+    game.time = game.time + 1;
+    // use the time and modulus to control various speeds on the same game
+    // loop
```

Comment:
This is a great way to control speed but it is a bit complex especially getting students to understand mod.  I've talked a a number of people through this concept high school level to adult and it's always a bit of a challenge.  To make it slightly easier we could create a helper function called isWholeNumber.  Then it would be something like isWholeNumber(game.time / frequency) or we could do shouldRun(game.time, frequency).